### PR TITLE
Enable KASAN on build with Clang

### PR DESCRIPTION
**Pull request by cahirwpz**
We'd like to switch to Clang exclusively and ditch support for GCC. Hence all Clang builds need to pass when compiled with KASAN and LOCKDEP. There're some minor differences between Clang and GCC wrt. routines they expect to be implemented by ASAN runtime.

### CODE REVIEW

File: sys/riscv/boot.c [hunk 1]
```diff
@@ -69,49 +69,30 @@
...
-extern char __kernel_end_pa[];
+extern char __text[];
+extern char __data[];
+extern char __ebss[];
+extern char __kernel_start[];
+extern char __kernel_end[];
 
-#define RISCV_VIRTADDR(x)                                                      \
-  ((vaddr_t)((paddr_t)(x)-KERNEL_PHYS) | KERNEL_SPACE_BEGIN)
+static __noreturn __used void riscv_boot(paddr_t dtb, paddr_t pde,
```
Comment (MichalBlk) [inline sys/riscv/boot.c:78]: `__used` can be removed.

File: sys/riscv/boot.c [hunk 2]
```diff
@@ -69,49 +69,30 @@
...
 /*
  * Virtual memory boot data.
  */
 
-#define __bss_boot_stack __section(".bss.boot_stack")
-
 /* NOTE: the boot stack is used before we switch out to `thread0`. */
-__bss_boot_stack static __used alignas(STACK_ALIGN) uint8_t
-  boot_stack[PAGESIZE];
+static __used alignas(STACK_ALIGN) uint8_t boot_stack[PAGESIZE];
```
Comment (MichalBlk) [inline sys/riscv/boot.c:95]: `__used` can be removed.

File: build/arch.mips.mk [hunk 3]
```diff
@@ -4,23 +4,23 @@
...
+  TARGET := mipsel-mimiker-elf
+  ABIFLAGS := -mips32r2 -EL -G 0
   ELFTYPE := elf32-littlemips
 endif
-ELFARCH := mips
 
 ASAN_SHADOW_OFFSET := 0xD8000000
 
 ifeq ($(KERNEL), 1)
   # Added to all files
```
[the remark this answers is not in the document]
Comment (cahirwpz) [reply build/arch.mips.mk:24]: Done!

File: build/arch.riscv.mk [hunk 4]
```diff
@@ -42,8 +41,13 @@ ifeq ($(BOARD), sifive_u)
   endif
 endif
 
-GCC_ABIFLAGS += -march=rv$(XLEN)$(EXT) -mabi=$(ABI) 
-CLANG_ABIFLAGS += -target $(TARGET) -march=rv$(XLEN)$(EXT) -mabi=$(ABI)
+ifeq ($(LLVM), 1)
+  TARGET := riscv$(XLEN)-linux-mimiker-elf
+  ABIFLAGS += -march=rv$(XLEN)$(EXT) -mabi=$(ABI)
```
[the remark this answers is not in the document]
Comment (cahirwpz) [reply build/arch.riscv.mk:46]: Done!

File: sys/riscv/boot.c [hunk 5]
```diff
@@ -69,49 +69,30 @@
...
+extern char __kernel_end[];
 
-#define RISCV_VIRTADDR(x)                                                      \
-  ((vaddr_t)((paddr_t)(x)-KERNEL_PHYS) | KERNEL_SPACE_BEGIN)
+static __noreturn __used void riscv_boot(paddr_t dtb, paddr_t pde,
+                                         paddr_t kern_end);
 
-#define KERNEL_VIRT_IMG_END align(RISCV_VIRTADDR(__ebss_pa), PAGESIZE)
-#define KERNEL_PHYS_IMG_END align((paddr_t)__ebss_pa, PAGESIZE)
+#define PHYSADDR(x) ((paddr_t)((vaddr_t)(x)-KERNEL_VIRT) | KERNEL_PHYS)
```
[the remark this answers is not in the document]
Comment (cahirwpz) [reply sys/riscv/boot.c:81]: Done!

File: sys/riscv/boot.c [hunk 6]
```diff
@@ -69,49 +69,30 @@
...
 #include <riscv/pmap.h>
 
-extern char __riscv_boot_pa[];
-extern char __boot_stack_pa[];
-extern char __kernel_start_pa[];
-extern char __text_pa[];
-extern char __data_pa[];
-extern char __ebss_pa[];
-extern char __kernel_end_pa[];
+extern char __text[];
```
[the remark this answers is not in the document]
Comment (cahirwpz) [reply sys/riscv/boot.c:72]: Done!

File: sys/riscv/boot.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (cahirwpz) [reply sys/riscv/boot.c:78]: Done!

Comment (cahirwpz) [reply sys/riscv/boot.c:78]: Done!

File: sys/riscv/boot.c [hunk 7]
```diff
@@ -264,10 +263,8 @@ extern void cpu_exception_handler(void);
 extern void *board_stack(paddr_t dtb_pa, void *dtb_va);
 extern void __noreturn board_init(void);
 
-#define __text_riscv_boot __section(".text.riscv_boot")
-
-__text_riscv_boot static __noreturn __used void
-riscv_boot(paddr_t dtb, paddr_t pde, paddr_t kern_end) {
+static __noreturn __used void riscv_boot(paddr_t dtb, paddr_t pde,
```
[the remark this answers is not in the document]
Comment (cahirwpz) [reply sys/riscv/boot.c:266]: Done!

File: sys/riscv/trap.c [hunk 8]
```diff
@@ -252,8 +252,8 @@ __no_profile void trap_handler(ctx_t *ctx) {
      * frame we consider situation to be critical and panic.
      * Hopefully sizeof(ctx_t) bytes of unallocated stack space will be enough
      * to display error message. */
-    register_t sp = __sp();
-    if ((sp & (PAGESIZE - 1)) < sizeof(ctx_t))
+    uintptr_t sp = __sp();
+    if (sp < (uintptr_t)td->td_kstack.stk_base + sizeof(ctx_t))
```
[the remark this answers is not in the document]
Comment (cahirwpz) [reply sys/riscv/trap.c:256]: This code works well on both 32-bit and 64-bit architectures. We can port it to AArch64 if needed.

**Pull request merged**